### PR TITLE
Remove virtual keyword from most functions in TensorImpl

### PR DESCRIPTION
**Pull request by yf225**
As part of the Variable/Tensor merge, we want to de-virtualize TensorImpl as much as possible, to achieve performance gain for common Tensor functions such as `numel()` / `sizes()` / `dim()`.

This supersedes https://github.com/pytorch/pytorch/pull/11259.

### CODE REVIEW

**Comment by gchanan** [comment]
benchmark?

**Review by gchanan — CHANGES_REQUESTED** [review]
there are a bunch of comments in TensorImpl.h about avoiding virtual dispatch for perf that aren't relevant anymore.

File: c10/core/TensorImpl.cpp [hunk 1]
```diff
@@ -93,20 +96,28 @@ void TensorImpl::release_resources() {
...
 int64_t TensorImpl::stride(int64_t d) const {
+  TORCH_CHECK(type_id_ != UndefinedTensorId() && layout() == kStrided, type_id_, " does not have stride");
   d = at::maybe_wrap_dim(d, dim(), false);
   return strides_[d];
 }
 
 TensorImpl* TensorImpl::maybe_zero_dim(bool condition_when_zero_dim) {
+  // We only allow this operation on dense tensors
+  if (!(has_storage() && layout() == kStrided)) {
+    TORCH_CHECK(condition_when_zero_dim == (dim() == 0),
```
Comment (gchanan) [inline c10/core/TensorImpl.cpp]: I know you just moved this, but I think this should be an INTERAL_CHECK; this isn't a public API.

File: c10/core/UndefinedTensorImpl.h [hunk 2]
```diff
@@ -17,14 +17,7 @@ struct C10_API UndefinedTensorImpl final : public TensorImpl {
 #endif
     return &_singleton;
   }
-  IntArrayRef sizes() const override;
-  IntArrayRef strides() const override;
-  int64_t size(int64_t d) const override;
-  int64_t stride(int64_t d) const override;
-  int64_t dim() const override;
   bool has_storage() const override;
```
Comment (gchanan) [inline c10/core/UndefinedTensorImpl.h]: why is has_storage still virtual?

File: c10/core/TensorImpl.cpp (hunk 1 above)
Comment (yf225) [reply c10/core/TensorImpl.cpp]: Fixed.

File: c10/core/UndefinedTensorImpl.h (hunk 2 above)
Comment (yf225) [reply c10/core/UndefinedTensorImpl.h]: I am thinking whether the derived types of TensorImpl is responsible for telling whether it has storage. An alternative is to explicitly check whether the current tensor is sparse or MKLDNN tensor, and if so return false in `has_storage()`.

Comment (yf225) [reply c10/core/UndefinedTensorImpl.h]: I changed it to explicitly check whether the current tensor is sparse or MKLDNN tensor, and if so return false in `has_storage()`.

File: c10/core/TensorImpl.cpp [hunk 3]
```diff
@@ -115,10 +126,12 @@ TensorImpl* TensorImpl::maybe_zero_dim(bool condition_when_zero_dim) {
 }
 
 bool TensorImpl::has_storage() const {
-  return storage_;
+  TORCH_CHECK(type_id_ != UndefinedTensorId(), type_id_, " does not have has_storage");
+  return !is_sparse() && !is_mkldnn() && storage_;
```
Comment (gchanan) [inline c10/core/TensorImpl.cpp]: isn't this just strided?  Do we ever have strided tensors where "storage_" is false?

File: c10/core/TensorImpl.cpp [hunk 4]
```diff
@@ -93,20 +96,28 @@ void TensorImpl::release_resources() {
...
 
 int64_t TensorImpl::stride(int64_t d) const {
+  TORCH_CHECK(type_id_ != UndefinedTensorId() && layout() == kStrided, type_id_, " does not have stride");
   d = at::maybe_wrap_dim(d, dim(), false);
   return strides_[d];
 }
 
 TensorImpl* TensorImpl::maybe_zero_dim(bool condition_when_zero_dim) {
+  // We only allow this operation on dense tensors
+  if (!(has_storage() && layout() == kStrided)) {
```
Comment (gchanan) [inline c10/core/TensorImpl.cpp]: same as before, is there a difference between has_storage and the layout being strided?

Comment (gchanan) [reply c10/core/TensorImpl.cpp]: I guess my main question is -- can we just has_storage -> storage_ != NULL as an equivalence for kStrided?  I don't know if that works with C2 tensors or not, though.

**Pull request closed without merging**